src/main/java/com/hotels/bdp/circustrain/bigquery/extraction/BigQueryDataExtractionManager.java

```diff
@@ -20,51 +20,130 @@
 import java.util.HashMap;
 import java.util.Map;
 
-import com.google.cloud.bigquery.Table;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
-import com.hotels.bdp.circustrain.api.CircusTrainException;
+import com.google.cloud.bigquery.Table;
+import com.google.common.annotations.VisibleForTesting;
 
 public class BigQueryDataExtractionManager {
 
-  private final Map<String, BigQueryExtractionData> cache = new HashMap<>();
+  @VisibleForTesting
+  static class TableWrapper {
+
+    private final Table table;
+
+    TableWrapper(Table table) {
+      this.table = table;
+    }
+
+    @Override
+    public String toString() {
+      return table.getTableId().getDataset() + "." + table.getTableId().getTable();
+    }
+
+    @Override
+    public int hashCode() {
+      return (makeKey(table.getTableId().getDataset(), table.getTableId().getTable())).hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (this == obj) {
+        return true;
+      }
+      if (obj == null) {
+        return false;
+      }
+      if (getClass() != obj.getClass()) {
+        return false;
+      }
+      TableWrapper other = (TableWrapper) obj;
+      Table otherTable = other.table;
+      String thisDatasetName = this.table.getTableId().getDataset();
+      String otherDatasetName = otherTable.getTableId().getDataset();
+      if (!thisDatasetName.equals(otherDatasetName)) {
+        return false;
+      }
+
+      String thisTableName = this.table.getTableId().getTable();
+      String otherTableName = otherTable.getTableId().getTable();
+      if (!thisTableName.equals(otherTableName)) {
+        return false;
+      }
+
+      return true;
+    }
+  }
+
+  private static final Logger log = LoggerFactory.getLogger(BigQueryDataExtractionManager.class);
+
   private final BigQueryDataExtractionService service;
 
+  private Map<TableWrapper, BigQueryExtractionData> cache = new HashMap<>();
```

Comment:
I'm also a bit confused as to why we need a cache?

Reply:
Hive MetastoreClient makes multiple calls to get a extracted data location for a table, each call needs to generate the same location, also the ordering of operations is weird, metastore client runs first on source end trying to find a location for the table, and this needs to be generated on the fly then located later so the extraction can occur after the fact in copier, this same location then needs to be referenced on replication success on failure at the end of each Circus Train job as well